contracts/exchange/contracts/src/MixinMatchOrders.sol

```diff
@@ -262,6 +355,237 @@ contract MixinMatchOrders is
         }
     }
 
+    /// @dev Match complementary orders that have a profitable spread.
+    ///      Each order is filled at their respective price point, and
+    ///      the matcher receives a profit denominated in the left maker asset.
+    ///      This is the reentrant version of `batchMatchOrders` and `batchMatchOrdersWithMaximalFill`.
+    /// @param leftOrders Set of orders with the same maker / taker asset.
+    /// @param rightOrders Set of orders to match against `leftOrders`
+    /// @param leftSignatures Proof that left orders were created by the left makers.
+    /// @param rightSignatures Proof that right orders were created by the right makers.
+    /// @param shouldMaximallyFillOrders A value that indicates whether or not the order matching
+    ///                        should be done with maximal fill.
+    /// @return batchMatchedFillResults Amounts filled and profit generated.
+    function _batchMatchOrders(
+        LibOrder.Order[] memory leftOrders,
+        LibOrder.Order[] memory rightOrders,
+        bytes[] memory leftSignatures,
+        bytes[] memory rightSignatures,
+        bool shouldMaximallyFillOrders
+    )
+        internal
+        returns (LibFillResults.BatchMatchedFillResults memory batchMatchedFillResults)
+    {
+        // Ensure that the left and right orders have nonzero lengths.
+        if (leftOrders.length == 0) {
+            _rrevert(BatchMatchOrdersError(BatchMatchOrdersErrorCodes.ZERO_LEFT_ORDERS));
+        }
+        if (rightOrders.length == 0) {
+            _rrevert(BatchMatchOrdersError(BatchMatchOrdersErrorCodes.ZERO_RIGHT_ORDERS));
+        }
+
+        // Ensure that the left and right arrays are compatible.
+        if (leftOrders.length != leftSignatures.length) {
+            _rrevert(BatchMatchOrdersError(BatchMatchOrdersErrorCodes.INCOMPATIBLE_LEFT_ORDERS));
+        }
+        if (rightOrders.length != rightSignatures.length) {
+            _rrevert(BatchMatchOrdersError(BatchMatchOrdersErrorCodes.INCOMPATIBLE_RIGHT_ORDERS));
+        }
+
+        // Without simulating all of the order matching, this program cannot know how many
+        // matches there will be. To ensure that batchMatchedFillResults has enough memory
+        // allocated for the left and the right side, we will allocate enough space for the
+        // maximum amount of matches (the left side length added to the right side length).
+        uint256 maxLength = leftOrders.length + rightOrders.length;
```

Comment:
Seems like the upper bound should be `left + right - 1`?

Reply:
You're correct. 

Proof: In the worst case scenario, every order except for one must be maximally filled, every match in the batch except for the last match (the last match doesn't matter) only maximally fills one order, and every order can be filled. One order is maximally filled during each match by the definition of our order matching strategy, so in the worst case `leftOrders.length + rightOrders.length - 1`.